First stab at updating token

lwasser opened the pull request:
closes #585

## Description

This is a stab at an automatic token refresh built into stravalib using environment variables if the user wishes to do so. it should NOT impact current functionality. 

My notes are below regarding how this works. Once we are happy with it I will add docs as well. 


This workflow assumes that you can access your refresh token and token 
expiration date. This information is created the first time that you 
authenticate a user with Strava. 

The basic authentication information you need:

* **access_token**: this token value expires after 6 hours
* **refresh_token**: this token is needed to refresh the access_token after it expires
* **expires_at**: this is the date/time when your access time expires and you need 
to refresh it

The access token that you initially retrieve can be used for the full period it's 
active. Once it expires, you will need to use the refresh token to refresh it so you 
can continue to make API calls. 

The expires_at value allows stravalib to check to see if your token is active or 
not before making API calls allowing stravalib ti automatically refresh your token 
when needed. 

How this works... 

1. Create the .env file with the client secret and id 
2. ake sure python-dotenv is installed (optional dependency)
3. Populate the .env file with the client secret and ID values:

CLIENT_SECRET=
CLIENT_ID=

4. the rest will be done for you!
## Type of change

Select the statement best describes this pull request.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] This is a documentation update
- [ ] This is a infrastructure update (docs, ci, etc)
- [ ] Other (please describe)

## Does your PR include tests

If you are fixing a bug or adding a feature, we appreciate (but do not require)
tests to support whatever fix of feature you're implementing.

- [x] Yes
- [ ] No, i'd like some help with tests
- [ ] This change doesn't require tests

I'M GOING TO WORK ON TESTS NEXT. right now this pr breaks everything but I'll slowly fix it. I wanted folks to see this approach in case there is feedback as I'm working!

## Did you include your contribution to the change log?

- [X] Yes, `changelog.md` is up-to-date.

all coming soon...

lwasser commented on README.md:
````diff
@@ -157,6 +153,42 @@ if time.time() > client.token_expires_at:
     expires_at = refresh_response["expires_at"]
 ```
 
+## Automatic token refresh
````
I think it would be nice to have a tiny quick start and then we can link to our docs so we aren't trying to maintain examples in two places!

lwasser replied:
Update: in #613 i just linked to the docs to avoid maintaining instructions in several places. so I'll rebase this pr after we review merge that other pr - Or I'll rebase that one against this if this gets merged first. but the goal is to not dump all docs into the readme rather keep them confined to the docs and link to them.

lwasser commented on src/stravalib/protocol.py:
```diff
@@ -8,11 +8,14 @@
 from urllib.parse import urlencode, urljoin, urlunsplit
 
 import requests
+from dotenv import load_dotenv
```
Note: this is a new dependency, but we could make it optional. Users can always bypass dotenv and just set envt variables themselves too.

lwasser commented:
Anyone is welcome to test this out. i have to fix all of the mypy and test issues next but the functionality (I think) works. but i'd love some folks to test!! 🚀  I've moved all messaging to logging rather than print statements following @jsamoocha feedback example decorator!

lwasser commented on src/stravalib/protocol.py:
```diff
@@ -89,6 +102,85 @@ def __init__(
+        client_id = os.environ.get("CLIENT_ID")
+        client_secret = os.environ.get("CLIENT_SECRET")
+
+        if client_id and client_secret:
```
I'm stuck here on typing and how to ensure client_id is an int. it is possible that a user could write the wrong value to their envt or env file. mypy however is giving me errors .

jsamoocha commented on src/stravalib/protocol.py:
```diff
@@ -89,6 +102,168 @@ def __init__(
+                " If you need .env support, please install `python-dotenv`."
+            )
+        # os.environ.get returns strings by default
+        client_id = int(os.environ.get("CLIENT_ID"))
+        client_secret = os.environ.get("CLIENT_SECRET")
```
Maybe rename these to `STRAVA_CLIENT_ID` and `STRAVA_CLIENT_SECRET`? In larger apps, there will be dozens of env variables.

lwasser replied:
Good call! renamed!

jsamoocha commented on src/stravalib/protocol.py:
```diff
@@ -89,6 +102,168 @@ def __init__(
+            token.
+
+        """
+        try:
+            # If the package is not typed, mypy is not happy, ignore
+            # https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
+            from dotenv import load_dotenv  # type: ignore
+
+            load_dotenv()
+            logging.info("Loaded .env file successfully.")
+        except ImportError:
+            logging.debug(
+                "python-dotenv is not installed. Skipping .env file loading."
+                " If you need .env support, please install `python-dotenv`."
+            )
```
I think (in hindsight?) we should entirely leave out any possible .env file processing. Larger apps would include `stravalib` as a dependency and have their own management of development/test/production environments and would use `load_dotenv` on a higher level.

For smaller hobby projects, we could add references to our documentation on how to use .env files + pydotenv (there are plenty of great tutorials online).

The only thing our code should do is assume these env variables are set in whichever way the user prefers.

lwasser replied:
That's a good point. I struggled with this because i didn't like adding an (optional) dependency. this will make the code much cleaner as we only need the os library to access envt variables. 

I have a doc locally where i'm building out the steps so I'll just update our tutorials accordingly and we can link to the dot file tutorials as well. It's pretty simple to setup and now I don't have to worry about tests for it!

jsamoocha commented on src/stravalib/client.py:
```diff
@@ -67,6 +67,8 @@ def __init__(
         rate_limiter: limiter.RateLimiter | None = None,
         requests_session: Session | None = None,
+        token_expires: int | None = None,
+        token_refresh: str | None = None,
```
Let's rename this throughout to `refresh_token` to retain consistency with Strava's naming and our naming of `access_token`.

lwasser replied:
I've done this throughout! Are you comfortable with keeping `token_expires` as that makes it a verb?

jsamoocha commented on src/stravalib/protocol.py:
```diff
@@ -89,6 +102,168 @@ def __init__(
+            The parsed JSON response.
+        """
+        # Try to refresh the token unless told to not try
+        if refresh_expired_token:
```
To prevent that infinite loop, could we check the URL at this point and ignore refreshing if the URL itself ends with `/oauth/token`? In that case, we could leave the existing `exchange_code_for_token()` method unchanged.

lwasser replied:
Ahhhh so much better. One less parameter as well in this method. Thanks!!

jsamoocha reviewed (commented):
Thanks! Almost there, I left a few comments for improvement.

lwasser commented:
I will come back to this and work more on tests and mypy. but the basic code is working and I've addressed all comments in the review so far! (I think?)

lwasser commented on src/stravalib/protocol.py:
```diff
@@ -89,6 +102,187 @@ def __init__(
+
+        # Only refresh token if we know the users' environment is setup
+        if "/oauth/token" not in url and credentials:
+            self.refresh_expired_token(credentials[0], credentials[1])
```
This is what i don't like. I wanted to unpack a tuple with client_id, client_secret. It would make the code more expressive. The alternative would be that the tuple contains a third value that is a bool for valid credentials? that would make that method a bit more complex but would make the code here more expressive.

jsamoocha replied:
If you move the check and extraction of credentials to `APIV3`s initializer, this should be no problem anymore.

lwasser replied:
it's much better! thank you for catching that. it just gets assigned to self now which makes everything simpler.

lwasser commented on README.md:
```diff
@@ -157,6 +153,42 @@ if time.time() > client.token_expires_at:
+
+To do this add the following values to your environment yourself.
+
+### Using the dot file approach
```
Remove this given we've adjusted the approach.

lwasser commented on pyproject.toml:
```diff
@@ -62,7 +62,10 @@ build = ["build"]
   "pytest-cov",
-  "responses"
+  # Run tests in parallel
+  "pytest-xdist",
```
this supports distributed processing if available.

lwasser commented on src/stravalib/client.py:
```diff
@@ -111,21 +121,63 @@ def access_token(self) -> str | None:
         """
         self.protocol.access_token = token_value
 
+    # TODO: if we go this route create a function to set up these attrs
+    @property
```
```suggestion
    @property
```

lwasser commented on src/stravalib/protocol.py:
```diff
@@ -89,6 +102,190 @@ def __init__(
             lambda _request_params, _method: None
         )
 
+    def _check_credentials(self) -> tuple[int, str] | None:
```
This is a new helper that checks credentials. This makes it easier to test and hopefully maintain and avoids adding a lot of logic to the refresh steps.

jsamoocha replied:
Great refactoring. We can simplify this by always returning a tuple of potentially nullables (so the return type is `tuple[int | None, str | None]`) and storing the result as instance variables of `APIV3` for later reuse. See the other comments.

lwasser commented on src/stravalib/protocol.py:
```diff
@@ -89,6 +102,190 @@ def __init__(
+            The parsed JSON response.
+        """
+
+        # If the request is a token request, skip, otherwise
+        # refresh token
+        # This feels sloppy but if check credentials is None then that means
+        # users environment isn't setup
```
```suggestion
        # Check if refresh token credentials are setup
```

lwasser commented on src/stravalib/protocol.py:
```diff
@@ -264,77 +471,6 @@ def resolve_url(self, url: str) -> str:
             )
         return url
 
-    def _request(
```
i didn't delete or modify much in this method. It looks that way because of the new methods and where I added them.

lwasser commented on src/stravalib/tests/unit/test_protocol_ApiV3_token_refresh.py:
```diff
@@ -0,0 +1,105 @@
+from stravalib.protocol import ApiV3
+
+
+@pytest.fixture
+def apiv3_instance():
+    """Fixture to create an ApiV3 instance for testing."""
+
+    return ApiV3()
```
```suggestion
```

moved to conftest

lwasser commented on src/stravalib/tests/unit/test_protocol_ApiV3_token_refresh.py:
```diff
@@ -0,0 +1,105 @@
+
+
+def test_request_skips_refresh_for_oauth_token(apiv3_instance):
+    """Test that _request skips refresh_expired_token for /oauth/token URL."""
```
I had to mock a lot here to isolate the single URL check. could I use our mock API object? i haven't dug into that yet.

lwasser commented on src/stravalib/tests/unit/test_refresh_token.py:
```diff
@@ -0,0 +1,123 @@
+"""A series of unit tests to ensure the refresh token operations
+work as expected whether the user has set things up or not.
+
+TODO:
+
+exchange_code_for_token
+* make sure that access token and token expires are set properly in the
+ - protocol.py:360-370 & 414-419 (how are these different?)
+
+ ## Client.py
+ * in client - make sure the access token, refresh value and expires at are all
+  set client.py:120-179
+* refresh access token: 258
+i think these will be integration tests to add to client?
+vs unit
```
```suggestion
```

lwasser commented on src/stravalib/tests/unit/test_refresh_token.py:
```diff
@@ -0,0 +1,123 @@
+    apiv3_instance.refresh_token = "dummy_refresh_token"
+
+    # Patch os get envt variables and populate envt vals
+    with patch.dict(
+        os.environ,
+        {"STRAVA_CLIENT_ID": "12345", "STRAVA_CLIENT_SECRET": "secret"},
+    ):
```
```suggestion
```

this context manager is no longer needed because it only is needed in the _check_credentials helper (I think)

lwasser commented on src/stravalib/tests/unit/test_refresh_token.py:
```diff
@@ -0,0 +1,123 @@
+            client_id=12345, client_secret="123secret"
+        )
+
+        # Check how many times refresh_access_token was called
+        assert apiv3_instance.refresh_access_token.call_count == expected_call
```
```suggestion
        assert apiv3_instance.refresh_access_token.call_count == expected_call
```

lwasser commented:
Ok this is ready for another review and I will pause working on things. 
✅ I added tests but am uncertain whether what I added is sufficient. I focused on unit tests that isolate specific behaviors. 
✅ I cleaned up the feature code following the review. 

All feedback welcome! 🎉

jsamoocha commented on src/stravalib/protocol.py:
```diff
@@ -203,62 +242,42 @@ def _token_expired(self) -> bool:
             logging.warning(
-                "Client ID and secret not found. You need to manually refresh"
-                " your token or set up environment variables."
+                "Please make sure you've set up your environment."
+                " I can't find your refresh_token"
```
This warning message may be confusing as the refresh token is not coming from the environment, and the warning may also be triggered if the refresh token is present but has not expired yet. We may need separate warnings for each case.

lwasser replied:
fixed - i cleaned that method up and added more explicit feedback if refresh_token isn't populated.

jsamoocha commented on src/stravalib/protocol.py:
```diff
@@ -131,18 +169,19 @@ def _request(
-        if refresh_expired_token:
-            self.refresh_expired_token()
+
+        # Check if refresh token credentials are setup
+        credentials = self._check_credentials()
```
It's great that you refactored this 👍

However, it may be better to move this check to `APIV3`s initializer. Otherwise, this check (and its associated warning/error messages) will be performed on every request.

lwasser replied:
Ok this has been moved to init. it's much better now. I'm really glad that you caught that. I'm refactoring tests a bit and I'll look at that comment above.

jsamoocha reviewed (changes requested):
Thanks for these changes! There is one remaining issue about the environment being checked for credentials on every request. It can be fixed by moving the check to the `APIV3` initializer.

lwasser commented on src/stravalib/protocol.py:
```diff
@@ -88,6 +103,175 @@ def __init__(
+                "Please make sure your STRAVA_CLIENT_ID is set in your environment."
+            )
+
+        if client_id and client_secret:
```
Populate credentials to self if they exist and are in the correct format.

lwasser commented on src/stravalib/tests/conftest.py:
```diff
@@ -18,3 +21,26 @@ def mock_strava_api():
+
+
+@pytest.fixture
+def mock_strava_env():
```
vscode will STORE environment variables like this that are mocked. This was breaking things and I couldn't figure out why. This fixture allows us to mock that piece by just calling it.

lwasser commented on src/stravalib/protocol.py:
```diff
@@ -74,12 +80,27 @@ def __init__(
+        self.access_token: str | None = access_token
+        self.token_expires: int | None = token_expires
+        self.refresh_token: str | None = refresh_token
+        self.client_id: int | None = None
```
This two vars as now populated by the method _check_credentials which avoids having to pass things around and perform continual checks!

lwasser commented on src/stravalib/protocol.py:
```diff
@@ -88,6 +109,189 @@ def __init__(
             lambda _request_params, _method: None
         )
+        # Check for credentials when initializing
+        self._check_credentials()
```
Runs on init... if credentials exist, populate to self.

lwasser commented on src/stravalib/protocol.py:
```diff
@@ -88,6 +109,189 @@ def __init__(
+        """
+
+        # Only refresh token if we know the users' environment is setup
+        if "/oauth/token" not in url and self.client_id and self.client_secret:
```
The only things changed in _request is this check

lwasser commented on src/stravalib/protocol.py:
```diff
@@ -88,6 +109,189 @@ def __init__(
+        if not self._token_expired():
+            return
+
+        # This should never be false, BUT mypy wants a reminder that these values
+        # are populated
```
Double check this - it made mypy happy but I wonder if we need test coverage or if this is just a mypy work around.

lwasser replied:
I don't see a way around adding these asserts to make mypy happy. but client_id should never be missing here as it's checked before this method is even run.

lwasser commented on noxfile.py:
```diff
@@ -5,7 +5,7 @@
 import nox
 
-nox.options.reuse_existing_virtualenvs = False
+nox.options.reuse_existing_virtualenvs = True
```
```suggestion
nox.options.reuse_existing_virtualenvs = False
```
mistakenly committed this. we want these to rebuild but I've been running tests over and over so there's no point in rebuilding  every time

lwasser commented on src/stravalib/protocol.py:
```diff
@@ -238,6 +445,12 @@ def refresh_access_token(
+
+        Notes
+        -----
+        This method is user facing so we don't populate client_id and
+        client_secret from self in case user wish to manually populate
+        those values.
```
```suggestion
        This method is user facing. Here, we don't populate client_id and
        client_secret from self; A user can call this method and refresh the token manually.
        those values.
```

lwasser commented on src/stravalib/protocol.py:
```diff
@@ -88,6 +109,189 @@ def __init__(
+            If all is setup properly, updates and resets the access_token
+            attribute. Otherwise it logs a warning
+        """
+        if not self.refresh_token:
```
This now explicetly provides the user with feedback if they want to use token_refresh. hopefully it's more clear than before!

lwasser commented:
Ok this is ready for another review when you have time. I've cleaned up the tests some and addressed the review comments ✨

jsamoocha commented on src/stravalib/protocol.py:
```diff
@@ -238,6 +445,12 @@ def refresh_access_token(
+        Notes
+        -----
+        This method is user facing. Here, we don't populate client_id and
+        client_secret from self; A user can call this method and refresh the token manually.
```
```suggestion
        client_secret from self; A user can call this method and refresh the token manually with
```

jsamoocha reviewed (approved):
Great stuff! I found a typo in a docstring somewhere; otherwise, this can be merged.

lwasser commented:
yay thanks! i'll merge it now. Should we push a release with the new features?

jsamoocha commented:
A release sounds good. It will allow me to reduce some complexity in code at my end :-)

lwasser commented:
awesome. ok i'll merge now. And i'll do a release in the next few days 🚀
